src/devices/video/upd7220.cpp

```diff
@@ -999,58 +985,150 @@ void upd7220_device::draw_rectangle(int x, int y)
 //  draw_char -
 //-------------------------------------------------
 
-void upd7220_device::draw_char(int x, int y)
+void upd7220_device::draw_char()
 {
-	int isize,psize;
-	uint16_t tile_data;
+	const int8_t dir_change[2][4] = {
+		{2, 2, -2, -2},
+		{1, 3, -3, -1}
+	};
+	const uint8_t type = (m_figs.m_figure_type & 0x10) >> 4;
 
-	LOG("uPD7220 char check: %d %d %02x %08x %d %d %02x\n",x,y,m_figs.m_dir,m_ead,m_figs.m_d,m_figs.m_dc,m_figs.m_figure_type);
 
-	/* QX10 may require upper bits for psize, VT240 requires the opposite */
-	isize = m_figs.m_d;
-	psize = m_figs.m_dc + 1;
+	LOG("uPD7220 char check: %08x %04x %02x %02x %d %d %02x\n",m_ead,m_mask,m_bitmap_mod,m_figs.m_dir,m_figs.m_d,m_figs.m_dc,m_figs.m_figure_type);
 
-	for(int pi = 0; pi < psize; pi++)
+	for (int i = 0, di = 0; i < (m_figs.m_dc + 1); ++i)
 	{
-		tile_data = (m_ra[15-(pi & 7)] << 8) | m_ra[15-(pi & 7)];
-		for(int pz = 0; pz <= m_gchr; pz++)
+		m_pattern = (m_ra[15 - (i & 7)] << 8) | m_ra[15 - (i & 7)];
+		for (int zdc = 0; zdc <= m_gchr; ++zdc, ++di)
 		{
-			int ii = 0, curpixel = 0;
-			if(pi & 1)
-			{
-				ii = isize - 1;
-				curpixel = (isize * (m_gchr + 1)) - 1;
-			}
-			while(pi & 1 ? ii >= 0 : ii < isize)
+			for (int j = 0; j < m_figs.m_d; ++j)
 			{
-				for(int iz = 0; iz <= m_gchr; iz++)
+				const uint16_t pattern = (di % 2) ? get_pattern(15-(j & 0xf)) : get_pattern((j & 0xf));
+				for (int zd = 0; zd <= m_gchr; ++zd)
 				{
-					draw_pixel(x + (curpixel * x_dir[m_figs.m_dir]), y + (curpixel * y_dir[m_figs.m_dir]), ii, tile_data);
-					if(pi & 1)
-						curpixel--;
-					else
-						curpixel++;
+					write_vram(0, m_bitmap_mod, pattern, m_mask);
+
+					if (j != (m_figs.m_d - 1) || zd != m_gchr)
+						next_pixel(m_figs.m_dir);
 				}
-				if(pi & 1)
-					ii--;
-				else
-					ii++;
 			}
-			if(m_figs.m_figure_type == 2)
+			m_figs.m_dir = (((uint16_t)m_figs.m_dir + dir_change[type][(di % 2) << 1]) & 7);
+			next_pixel(m_figs.m_dir);
+			m_figs.m_dir = (((uint16_t)m_figs.m_dir + dir_change[type][((di % 2) << 1)+1]) & 7);
+		}
+	}
+}
+
+
+//-------------------------------------------------
+//  helper functions used to rotate a uint16_t
+//-------------------------------------------------
+
+static inline uint32_t rotate_right (uint16_t value)
+{
+	return (value>>1) | (value<<( (-1) & 0x0f ));
+}
+
+static inline uint32_t rotate_left (uint16_t value)
+{
+	return (value<<1) | (value>>( (-1) & 0x0f ));
+}
```

Comment:
Any reason these are returning `uint32_t`?  Am I missing somewhere it’s used to do a 17-bit rotate, or is it always assigned to a `uint16_t`?  They should probably be `constexpr`, and even if there’s code elsewhere in this file using `func_name (...)` (i.e. space between function name and opening parenthesis of formal parameter list), I really don’t want it spreading – no space between function name and opening parenthesis of formal parameter list is something we pretty much agree on all through MAME.

Reply:
Yeah that was a mistake on my part it should return a uint16_t since its just rotateing across 16 bits. I'm not sure why you say it should be a constexpr though? My understanding of constexpr is that it indicates something that can be evaluated at compile time, but this function is passed a variable value that wouldn't be known at compile time.

Reply:
`constexpr` means it’s a pure function – the result depends only on the inputs, there are no visible side effects, the result doesn’t depend on global state.  You can use `constexpr` functions with constant inputs in places where a constant is required, but there’s no requirement that you *only* use it in constant context.  It also implies internal linkage (i.e. has a similar effect to `inline`).

I mean, the `BIT` helper function and its friends are all `constexpr` and we use them on values that aren’t known at compile time all the time: https://github.com/mamedev/mame/blob/mame0234/src/lib/util/coretmpl.h#L584